### PR TITLE
initial leader election in place

### PR DESCRIPTION
**Pull request by depombo**

### CODE REVIEW

File: src/main/java/com/uber/kafkaSpraynozzle/KafkaSpraynozzle.java [hunk 1]
```diff
@@ -42,13 +47,39 @@ public static void main(String[] args) {
...
+        boolean buffering = spraynozzleArgs.getBuffering();
+        String hostName = getHostNameFromInetAddress();
+        if (isStringEmpty(hostName)) {
+            hostName = getHostNameFromFile();
+        }
+
+        if (!isStringEmpty(leaderLatchZkPath) && !isStringEmpty(counterZkPath)) {
+            System.out.println("Performing leader election through zookeeper and picking leader that will proceed.");
+            //use same zk as kafka
+            SpraynozzleLeaderLatch curatorClient = new SpraynozzleLeaderLatch(zk, leaderLatchZkPath, counterZkPath, "spraynozzle-" + hostName);
```
Comment (dfellis) [inline src/main/java/com/uber/kafkaSpraynozzle/KafkaSpraynozzle.java]: Can we include the topic name and cleaned URL, too? There may be more than one spraynozzle sitting on the same box, though part of a different cluster.

File: src/main/java/com/uber/kafkaSpraynozzle/KafkaSpraynozzle.java [hunk 2]
```diff
@@ -42,13 +47,40 @@ public static void main(String[] args) {
...
+        if (isStringEmpty(hostName)) {
+            hostName = getHostNameFromFile();
+        }
+
+        if (spraynozzleHA) {
+            System.out.println("Performing leader election through zookeeper and picking leader that will proceed.");
+            //use same zk as kafka
+            String zkLeaderLatchFolderPath = "/consumers/kafka_spraynozzle_leader_latch";
+            String zkLeaderElectionFolderPath = "/consumers/kafka_spraynozzle_leader_elections";
+            SpraynozzleLeaderLatch curatorClient = new SpraynozzleLeaderLatch(zk, zkLeaderLatchFolderPath, zkLeaderElectionFolderPath, "spraynozzle-" + hostName);
```
Comment (dfellis) [inline src/main/java/com/uber/kafkaSpraynozzle/KafkaSpraynozzle.java]: My comment seems to have disappeared. We'll need to include the topic name and cleaned url for the leader election, as well, to allow multiple rings of spraynozzles to coexist on the same hosts.

This is probably also true of the leader latch and leader election paths above. Multiple simultaneous clusters could be connected to the same zookeeper, so they need to be distinct.

**Comment by dfellis** [comment]
Just one comment on this diff that I think needs fixing. This is awesome work! :)

And we can improve this in the future to deal with network partitions, as well, but this is a million times better than the current situation. :trophy:

**Comment by depombo** [comment]
I do it gladly, David! Thanks for catching that. I have modified zk paths to include host, topic(s) and cleanedUrl and the spraynozzle identify within it with a UUID.

**Comment by dfellis** [comment]
And we finally got around to testing this, and it works well! Merging. :)

**Pull request merged**